### PR TITLE
ETS-525 Typescript conversion

### PR DESCRIPTION
**Pull request by hitankar-freshworks**
- [x] Conversion JS to Ts
- [x] CommonJs to ES imports
- [x] Validation fixes
- [x] Updated linting
- [x] Updated run and build commands.
- [x] Running in docker
- [x] Tests
- [x] Deployment changes

Note: Sorry for the big change but this cannot be avoided. When checking feel free to ignore the files that implement logic. They are identical except for small type fixes.

### CODE REVIEW

File: server/babel.config.js [hunk 1]
```diff
@@ -0,0 +1,7 @@
+module.exports = {
```
Comment (vyasworks) [inline server/babel.config.js:1]: do we need babel for backend?

Comment (hitankar-freshworks) [reply server/babel.config.js:1]: Unfortunately, Jest requires babel when using TypeScript [here](https://jestjs.io/docs/getting-started#using-typescript)

File: server/package.json [hunk 2]
```diff
@@ -5,9 +5,10 @@
   "author": "",
   "license": "ISC",
   "scripts": {
-    "db:seed": "node db/scripts/seed.js",
-    "watch": "NODE_ENV=development nodemon main.js",
-    "start": "NODE_ENV=production node main.js",
+    "db:seed": "tsc db/scripts/seed.ts",
+    "watch": "NODE_ENV=development nodemon main.ts",
+    "start": "NODE_ENV=production npm run build && node dist/main.js",
```
Comment (sunghwan-park-fw) [inline server/package.json:10]: Don't we need to change lambda entrypoints? I don't know where they are.

Comment (hitankar-freshworks) [reply server/package.json:10]: yeah, I am working on that but I am trying to figure out how to resolve the issue with `custom_modules` first.

[the remark this answers is not in the document]
Comment (hitankar-freshworks) [reply server/package.json:10]: updated to remove yarn

File: Makefile [hunk 3]
```diff
@@ -97,14 +97,16 @@ build-lambdas:
...
+	npm install --prefix server
 	npm install --production --prefix server/lambda/layer/common/nodejs/custom_modules
 	npm install --production --prefix server/lambda/layer/common/nodejs
 
 	@echo -e "\n\n\n============================="
 	@echo "Copy backend to temp folder to start build"
 	@echo "============================="
-	cp -r ./server ./terraform/build/server
+	npm run build --prefix server
+	rm -rf server/node_modules && npm install --production --prefix server
```
Comment (sunghwan-park-fw) [inline Makefile:108]: we install at line 100, then why do we install again?

File: Makefile [hunk 4]
```diff
@@ -97,14 +97,16 @@ build-lambdas:
 	@echo -e "\n\n\n============================="
 	@echo "Install Node Modules in all packages"
 	@echo "============================="
-	npm install --production --prefix server
+	npm install --prefix server
 	npm install --production --prefix server/lambda/layer/common/nodejs/custom_modules
```
Comment (sunghwan-park-fw) [inline Makefile:101]: custom_modules has no package definition. I doubt this is effective?

**Review by sunghwan-park-fw — APPROVED** [review]

File: Makefile (hunk 3 above)
Comment (hitankar-freshworks) [reply Makefile:108]: well, we don't want to include devDependencies in the lambda zip. But we need to do a full install in order to run the build.

File: Makefile (hunk 4 above)
Comment (hitankar-freshworks) [reply Makefile]: It does have package.json file if that is what you mean.

**Review by sunghwan-park-fw — APPROVED** [review]

**Pull request merged**